### DISCUSSION TITLE
Unable to accomplish a distributed installation

### DISCUSSION THREAD

**Discussion by clopmz** [telekom-security/tpotce#1153] [Q&A]
Hi all,

I am trying to accomplish a distributed install with T-PotCE release 22.04 ... When I try to execute deploy.sh script from the hive_sensor, the following error is returned:
```
###################################
# Deploy T-Pot Logs to T-Pot Hive #
###################################

[c] - Continue deplyoment
[q] - Abort and exit

Your choice: c

### Please provide data from your T-Pot Hive installation.
### This usually is the one running the 'T-Pot Hive' type.
### You will be needing the OS user (typically 'tsec'), the users' password and the IP / FQDN.
### Do not worry, the password will not be persisted!

Username: root
Password:
IP / FQDN: 172.22.55.8

##############################
# Generating ssh keyfile ... #
##############################

Generating public/private rsa key pair.
Your identification has been saved in /data/elk/logstash/educationalflower
Your public key has been saved in /data/elk/logstash/educationalflower.pub
The key fingerprint is:
SHA256:XO75nT+IC/18xtpCiTmCwW4rdJo1BbKy2aErZw5opbs educationalflower
The key's randomart image is:
+---[RSA 3072]----+
|                 |
|     . .         |
|      + . .      |
|   . o + +       |
|   .* o S . o .  |
|. o+ o B o.= o   |
|.+  o * o.+.+ o  |
|..o+ + .  .oo+.* |
| E*.  .    .o+Boo|
+----[SHA256]-----+

###########################################################
# Writing config to /data/elk/logstash/ls_environment.    #
# If you make changes to this file, you need to reboot or #
# run /opt/tpot/bin/updateip.sh.                          #
###########################################################

MY_TPOT_TYPE=SENSOR
MY_SENSOR_PRIVATEKEYFILE=/data/elk/logstash/educationalflower
MY_HIVE_USERNAME=educationalflower
MY_HIVE_IP=172.22.55.8

############################################
# Checking for T-Pot Hive availability ... #
############################################


#########################
# T-Pot Hive available! #
#########################


######################################################
# T-Pot Hive tunnel test FAILED!                     #
# Tunneled port tcp/64305 unreachable on T-Pot Hive. #
# Aborting.                                          #
######################################################

kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
```
hive_sensor can access to hive via SSH in port 64295.

What am I doing wrong?

**Comment by t3chn0m4g3** [comment]
`root` is not permitted to login remotely. You need to use a user with SSH login permission.

**Reply by clopmz** [reply]
Uhmm ... I have enabled "PermitRootLogin yes" in hive's sshd_config file to allow this initial contact from hive_sensor ... Do I need to create a normal user in hive server only?

**Reply by t3chn0m4g3** [reply]
I doubt this will work with `root` and it is not recommended.

**Reply by clopmz** [reply]
Ok, thanks ... I will create a normal user and test it.

**Reply by clopmz** [reply]
Uhmm ... Same result:

Your choice: c

### Please provide data from your T-Pot Hive installation.
### This usually is the one running the 'T-Pot Hive' type.
### You will be needing the OS user (typically 'tsec'), the users' password and the IP / FQDN.
### Do not worry, the password will not be persisted!

Username: sysop
Password:
IP / FQDN: 172.22.55.8

##############################
# Generating ssh keyfile ... #
##############################

Generating public/private rsa key pair.
Your identification has been saved in /data/elk/logstash/educationalflower
Your public key has been saved in /data/elk/logstash/educationalflower.pub
The key fingerprint is:
SHA256:1+5yquo/YVYtfWtNmjzmhI+3GGgYnC9xwz2z7h9Zs8A educationalflower
The key's randomart image is:
+---[RSA 3072]----+
|                 |
|                 |
|           o     |
|       . ooo+ . .|
|        S.=.=E B.|
|        +B +.+@o+|
|       oo.+ +Boo |
|        .o.ooo+. |
|     .oooo.==oo. |
+----[SHA256]-----+

###########################################################
# Writing config to /data/elk/logstash/ls_environment.    #
# If you make changes to this file, you need to reboot or #
# run /opt/tpot/bin/updateip.sh.                          #
###########################################################

MY_TPOT_TYPE=SENSOR
MY_SENSOR_PRIVATEKEYFILE=/data/elk/logstash/educationalflower
MY_HIVE_USERNAME=educationalflower
MY_HIVE_IP=172.22.55.8

############################################
# Checking for T-Pot Hive availability ... #
############################################


#########################
# T-Pot Hive available! #
#########################


######################################################
# T-Pot Hive tunnel test FAILED!                     #
# Tunneled port tcp/64305 unreachable on T-Pot Hive. #
# Aborting.                                          #
######################################################

kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]

**Reply by t3chn0m4g3** [reply]
Were you using the Post Install method? If yes, please try the ISO, otherwise retrace all steps from `deploy.sh` to see where / what the culprit is.

**Reply by clopmz** [reply]
Yes, I have installed both tpotce nodes in a Debian 11 clean install ..

**Comment by clopmz** [comment]
Uhmm .. It seems a bug. I have enabled "-x" flag in deploy.sh script and returuns this error:
```
+ tee /data/elk/logstash/ls_environment
MY_TPOT_TYPE=SENSOR
MY_SENSOR_PRIVATEKEYFILE=/data/elk/logstash/educationalflower
MY_HIVE_USERNAME=educationalflower
MY_HIVE_IP=172.22.55.8
+ fuCHECK_HIVE
+ echo

+ echo '############################################'
############################################
+ echo '# Checking for T-Pot Hive availability ... #'
# Checking for T-Pot Hive availability ... #
+ echo '############################################'
############################################
+ echo

+ sshpass -e ssh -4 -t -l sysop -p 64295 -f -N -L64305:127.0.0.1:64305 172.22.55.8 -o StrictHostKeyChecking=no
+ '[' 0 -eq 0 ']'
+ echo

+ echo '#########################'
#########################
+ echo '# T-Pot Hive available! #'
# T-Pot Hive available! #
+ echo '#########################'
#########################
+ echo

++ curl -s http://127.0.0.1:64305
+ myHIVE_OK=
+ '[' '' == ok ']'
+ echo

+ echo '######################################################'
######################################################
+ echo '# T-Pot Hive tunnel test FAILED!                     #'
# T-Pot Hive tunnel test FAILED!                     #
+ echo '# Tunneled port tcp/64305 unreachable on T-Pot Hive. #'
# Tunneled port tcp/64305 unreachable on T-Pot Hive. #
+ echo '# Aborting.                                          #'
# Aborting.                                          #
+ echo '######################################################'
######################################################
+ echo

++ pidof ssh
+ kill -9
kill: usage: kill [-s sigspec | -n signum | -sigspec] pid | jobspec ... or kill -l [sigspec]
+ rm /data/elk/logstash/educationalflower.pub
+ rm /data/elk/logstash/educationalflower
+ rm /data/elk/logstash/ls_environment
+ exit 1

But running commands in another console, it works:

[root@educationalflower:~]# ssh -4 -t -l sysop -p 64295 -f -N -L64305:127.0.0.1:64305 172.22.55.8
sysop@172.22.55.8's password:
[root@educationalflower:~]# curl -s http://127.0.0.1:64305
ok[root@educationalflower:~]#
```
As you can see, curl works ... but not from deploy.sh script ...

**Reply by t3chn0m4g3** [reply]
I cannot reproduce this with the ISO installs I am currently having.

**Reply by t3chn0m4g3** [reply]
Fresh ISO deployments work perfectly fine as well.

```
[root@surroundingcrotch:~]# bash -x deploy.sh
+ fuGOT_ROOT
+ echo

+ echo -n '### Checking for root: '
### Checking for root: ++ whoami
+ '[' root '!=' root ']'
+ echo '[ OK ]'
[ OK ]
+ echo

+ echo '#################################'
#################################
+ echo '# Ship T-Pot Logs to T-Pot Hive #'
# Ship T-Pot Logs to T-Pot Hive #
+ echo '#################################'
#################################
+ echo

+ echo 'If you already have a T-Pot Hive installation running and'
If you already have a T-Pot Hive installation running and
+ echo 'this T-Pot installation is running the type "Pot" the'
this T-Pot installation is running the type "Pot" the
+ echo 'script will automagically setup this T-Pot to ship and'
script will automagically setup this T-Pot to ship and
+ echo 'prepare the Hive to receive logs from this T-Pot.'
prepare the Hive to receive logs from this T-Pot.
+ echo

+ echo

+ echo '###################################'
###################################
+ echo '# Deploy T-Pot Logs to T-Pot Hive #'
# Deploy T-Pot Logs to T-Pot Hive #
+ echo '###################################'
###################################
+ echo

+ echo '[c] - Continue deplyoment'
[c] - Continue deplyoment
+ echo '[q] - Abort and exit'
[q] - Abort and exit
+ echo

+ '[' 1 '!=' 2 ']'
+ read -s -n 1 -p 'Your choice: ' mySELECT
Your choice: + echo c
c
+ case "$mySELECT" in
+ fuGET_DEPLOY_DATA
+ echo

+ echo '### Please provide data from your T-Pot Hive installation.'
### Please provide data from your T-Pot Hive installation.
+ echo '### This usually is the one running the '\''T-Pot Hive'\'' type.'
### This usually is the one running the 'T-Pot Hive' type.
+ echo '### You will be needing the OS user (typically '\''tsec'\''), the users'\'' password and the IP / FQDN.'
### You will be needing the OS user (typically 'tsec'), the users' password and the IP / FQDN.
+ echo '### Do not worry, the password will not be persisted!'
### Do not worry, the password will not be persisted!
+ echo

+ read -p 'Username: ' MY_TPOT_USERNAME
Username: tsec
+ read -s -p 'Password: ' SSHPASS
Password: + echo

+ export SSHPASS
+ read -p 'IP / FQDN: ' MY_HIVE_IP
IP / FQDN: 172.20.254.154
++ hostname
+ MY_HIVE_USERNAME=surroundingcrotch
+ MY_TPOT_TYPE=SENSOR
+ MY_LS_ENVCONFIGFILE=/data/elk/logstash/ls_environment
+ MY_SENSOR_PUBLICKEYFILE=/data/elk/logstash/surroundingcrotch.pub
+ MY_SENSOR_PRIVATEKEYFILE=/data/elk/logstash/surroundingcrotch
+ '[' -s /data/elk/logstash/surroundingcrotch ']'
+ '[' -s /data/elk/logstash/surroundingcrotch.pub ']'
+ echo

+ echo '##############################'
##############################
+ echo '# Generating ssh keyfile ... #'
# Generating ssh keyfile ... #
+ echo '##############################'
##############################
+ echo

+ mkdir -p /data/elk/logstash
+ ssh-keygen -f /data/elk/logstash/surroundingcrotch -N '' -C surroundingcrotch
Generating public/private rsa key pair.
Your identification has been saved in /data/elk/logstash/surroundingcrotch
Your public key has been saved in /data/elk/logstash/surroundingcrotch.pub
The key fingerprint is:
SHA256:LUrkVE8JOSRDPL/Olteil+ZJbh3L2liji8OrrjBRiuI surroundingcrotch
The key's randomart image is:
+---[RSA 3072]----+
|     o+.+o..     |
|      o+oo.      |
|    . oo ..      |
| . o +  ..       |
|o o   o S..      |
|o  . . ...  .    |
| Eo   .+ ..=oo   |
|   o    B+B*=.   |
|    .ooo+XO+.    |
+----[SHA256]-----+
++ cat /data/elk/logstash/surroundingcrotch.pub
+ MY_SENSOR_PUBLICKEY='ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABgQDoyikJyMHfiM0mY2/+rHHZkaIgo92olFGGJ/cjDiBqMaSwA2+fmv4+gVtjku+zKyswIKBztUMy8Z3iHFW6DQY9c7794sl4J3ku8QUmYmqri7XGJ0JTIq/RcaW9kIfFT2O3iPbs46h4CTVUTJr98Pjp4ZUVStUgt8jQ8Q0+iW40EbWH7069ewPCGhZkrLw2HNCLlc3xiz1C9BHugQEuDduSCQBEzGIWnlhG85piXCzaEyZAqQGKq1H3OHvjuJZSB5rpmpLVgj2LdlOD4AUC8i3pBfLN5dubADwnonfnj316Y9wYDO141tEn7mxfxU+zYzA+e6G7Nr7CqOIs0+8pG5+VeIkwmyeH/EeurU0iNJoqjgScJYfzILC83rtZJzaa38BsdQJ3L11IFJihxbPlqpTIsR+o5nPd4F3zD40VTaSWDsibqKKmdOj7oStQtGDQZa2DX7QPZEx4QVtm3IA1AhU4X/UjXDHZgOSYwrpdPcTKbo416rsJG3BaXnhr94qRWj0= surroundingcrotch'
+ echo

+ echo '###########################################################'
###########################################################
+ echo '# Writing config to /data/elk/logstash/ls_environment.    #'
# Writing config to /data/elk/logstash/ls_environment.    #
+ echo '# If you make changes to this file, you need to reboot or #'
# If you make changes to this file, you need to reboot or #
+ echo '# run /opt/tpot/bin/updateip.sh.                          #'
# run /opt/tpot/bin/updateip.sh.                          #
+ echo '###########################################################'
###########################################################
+ echo

+ tee /data/elk/logstash/ls_environment
MY_TPOT_TYPE=SENSOR
MY_SENSOR_PRIVATEKEYFILE=/data/elk/logstash/surroundingcrotch
MY_HIVE_USERNAME=surroundingcrotch
MY_HIVE_IP=172.20.254.154
+ fuCHECK_HIVE
+ echo

+ echo '############################################'
############################################
+ echo '# Checking for T-Pot Hive availability ... #'
# Checking for T-Pot Hive availability ... #
+ echo '############################################'
############################################
+ echo

+ sshpass -e ssh -4 -t -l tsec -p 64295 -f -N -L64305:127.0.0.1:64305 172.20.254.154 -o StrictHostKeyChecking=no
+ '[' 0 -eq 0 ']'
+ echo

+ echo '#########################'
#########################
+ echo '# T-Pot Hive available! #'
# T-Pot Hive available! #
+ echo '#########################'
#########################
+ echo

++ curl -s http://127.0.0.1:64305
+ myHIVE_OK=ok
+ '[' ok == ok ']'
+ echo

+ echo '##############################'
##############################
+ echo '# T-Pot Hive tunnel test OK! #'
# T-Pot Hive tunnel test OK! #
+ echo '##############################'
##############################
+ echo

++ pidof ssh
+ kill -9 11787
+ fuDEPLOY_SENSOR
+ echo

+ echo '###############################'
###############################
+ echo '# Deploying to T-Pot Hive ... #'
# Deploying to T-Pot Hive ... #
+ echo '###############################'
###############################
+ echo

+ sshpass -e ssh -4 -t -T -l tsec -p 64295 172.20.254.154
Linux digitaldictionary 5.10.0-17-arm64 #1 SMP Debian 5.10.136-1 (2022-08-13) aarch64
[sudo] password for tsec: + echo

+ echo '###########################'
###########################
+ echo '# Done. Please reboot ... #'
# Done. Please reboot ... #
+ echo '###########################'
###########################
+ echo

+ exit 0
```

**Reply by clopmz** [reply]
Ok ... tested and it works ... Many thanks for your help @t3chn0m4g3 

Only last question: how can I change passwords for tsec user and web admin?

**Reply by t3chn0m4g3** [reply]
Perfect. You can find the procedure in the [documentation](https://github.com/telekom-security/tpotce#add-users-to-nginx-t-pot-webui).

**Reply by clopmz** [reply]
Ok. Done

**Comment by Nuke492** [comment]
Hey im running into the same issue? any fix for this error? using the ISO method.

**Comment by lullu57** [comment]
Hi, I am running the Distributed version of TPot hosted on 2 bridged networking VMs. They are fresh tpot ISO downloads. The required ports for all the honey pots have been opened and directed to the sensor IP as required. When trying to run the deploy.sh script, using the standard tsec tpot user, the final connectivity test is failing:

T-Pot Hive tunnel test FAILED!
Tunneled port tcp/64305 unreachable on T-Pot Hive.

Running netstat -tulpn on the Hive shows that it is listening over port 64305 on the normal local address (127.0.0.1). Could you guide me as to how to troubleshoot this issue given that I am not using the root user, and I am using clean ISO downloads?

**Reply by t3chn0m4g3** [reply]
Please provide the debugs when running the script.

**Reply by lullu57** [reply]
![image](https://user-images.githubusercontent.com/79207487/196475401-c2935ec7-faae-4fc6-9e1b-d1b25e347f97.png)
![image](https://user-images.githubusercontent.com/79207487/196475574-8d1d1c4d-11ac-4b4f-a0d5-0ce14e2f5ae5.png)
Does this suffice?

**Reply by t3chn0m4g3** [reply]
No, please check the prev. [posts](https://github.com/telekom-security/tpotce/discussions/1153#discussioncomment-3415455).

**Reply by lullu57** [reply]
![image](https://user-images.githubusercontent.com/79207487/196625336-5610b6cf-a252-4f81-909d-b70731fe04a3.png)

![image](https://user-images.githubusercontent.com/79207487/196624833-a14c321c-27be-46b7-8ed4-c8f6acd7c64e.png)
![image](https://user-images.githubusercontent.com/79207487/196625555-679f0a6d-d11d-44c6-845f-b05cf4c5f247.png)
![image](https://user-images.githubusercontent.com/79207487/196625763-f0db4bfd-68a4-48be-8106-16609c268bf0.png)
![image](https://user-images.githubusercontent.com/79207487/196624961-aee1a106-d8b6-4d03-8ade-4c9422529061.png)

Does this work?

**Reply by t3chn0m4g3** [reply]
Thanks, unfortunately it is the same reason as mentioned above, and I cannot reproduce it. Neither with ISO nor Post Install method.

For some reason the ...
```
++ curl -s http://127.0.0.1:64305
+ myHIVE_OK=
+ '[' '' == ok ']'
+ echo
```
... tunnel test fails while trying to reach the remote endpoint via SSH tunnel, which usually is an indicator that some ports / service are unavailable, but this clearly is not the case here. Since I cannot reproduce it, maybe it is a timing issue. Can you add a `sleep 10` before the `curl` command to rule this out. If that is the case I can wait for the availability and break after a certain amount of time to avoid this.

**Reply by lullu57** [reply]
Hi I tried the above and still didn't work. Tried my luck with a fresh install again for both the hive and the sensors and connection worked this time. Not sure what the issue was before. Thankyou for your help. :)

**Reply by t3chn0m4g3** [reply]
Thanks for the feedback, I hope someone finds a common connection so I can finally reproduce this.

**Comment by NetManDE** [comment]
I created a pull request to fix this issue, which did it for me locally: https://github.com/telekom-security/tpotce/pull/1209

SSHPASS is only used in the Deploy Sensor function, where it deploys the SSH Key. Without a deployed SSH Key it will fail, as in the check Hive function it expects a cert based authentication, as no Password is being provided

**Reply by lullu57** [reply]
Thank you, hopefully it will fix the issue for everyone :)

**Comment by nicolasRamirez98** [comment]
Hello everyone, sorry if my English is not very good, I am using a translator.

I think I found the bug when deploying the sensor and the main hive. Basically I did the following:
1. parsing the script "deploy.sh", I find the line:
`sshpass -e ssh -4 -t -l "$MY_TPOT_USERNAME" -p 64295 -f -N -L64305:127.0.0.1:64305 "$MY_HIVE_IP" -o "StrictHostKeyChecking=no"`

![1](https://user-images.githubusercontent.com/46548508/221653464-8854d9b1-b718-4a41-93ff-46970611b81a.PNG)

This command allows to establish a tunnel between the two machines, what it does is that if in the sensor a query is executed to port 64305, it redirects to the same port of the hive machine. 

2. I do not know exactly why the `sshpass` command does not work well, however I think it may be when you try to export the environment variable SSHPASS, in the following line.
![2](https://user-images.githubusercontent.com/46548508/221653964-667fa4cc-e13f-45c8-992a-07e26e2b4d33.PNG)

3. Well, I did the following, seeing that the sshpass command was not working correctly.
- First run the sshpass command, to rule out any events.
- Then I just ran the ssh command with the related parameters in the script and it ran without any problems, as you can see below
![3](https://user-images.githubusercontent.com/46548508/221655227-460451f4-31a4-48e0-907d-40fa9b9f685e.PNG)
    You can see, that with red color I tried to explain that I had the error which we tried to solve.
With green color, I show the sshpass command with a small variation, which instead of the -e parameter I changed it to -p 'qwerty' which forces the password, instead of taking it from an environment variable called "SSHPASS".
Later I show with cyan color that after executing the ssh command and inserting the password, it already shows me the port being used.

- It finally worked by running the deploy.sh script again.
![4](https://user-images.githubusercontent.com/46548508/221656324-295e5ef7-b1de-41ec-b7af-55cce33937f7.PNG)

Additionally I have to say that I executed previous commands for the diagnostic, such as: `iptables` to open port 64305, `netstat` and `ss` to check the connections, and finally I studied the ssh documentation with: `man ssh`, this to understand what the parameters in the command do.

Please let me know if this solution worked for you, I will gladly read and help in any way I can.

**Reply by t3chn0m4g3** [reply]
It seems, as previously established, the reason is the `kill` command which is unavailable in some conditions. There seem to be specific conditions necessary for reproduction which do not seem to occur in our setup, so I cannot reproduce this.

**Reply by nicolasRamirez98** [reply]
Ok, when you say that about your conditions, what exactly do you mean, because, I am installing the two machines with the latest iso version, in a vmware esxi environment, and no previous packages installed or other things. Also both machines are in layer 2 Hive ip: 10.42.6.3 and the sensor ip: 10.42.6.4, so the communication is direct and does not go through firewalls or routers.

Thanks for you reply. :)

**Reply by t3chn0m4g3** [reply]
I have tested this with the iso and post install methods during development, after this discussion started and tried numerous times to reproduce this within VMWare Fusion, UTM, VirtualBox, Parallels and KVM even. Everything works perfectly fine as outlined in my previous posts.

**Comment by halfliffe** [comment]
I encountered the same error.
When I try to run deploy.sh on a clean installation of HIVE_SENSOR, I get the error:

![изображение](https://github.com/telekom-security/tpotce/assets/61053369/b1fa4506-1f7a-4747-bfa4-55fc76ff5eca)

Tell me how to fix this?
